source/georef/path_finder.cpp

```diff
@@ -0,0 +1,517 @@
+    auto target_dist = distances[target[target_e]] + crow_fly_duration(target.distances[target_e]);
+
+    if (target_dist < source_dist)
+        return {target_dist, target_e};
+
+    return {source_dist, source_e};
+}
+
+Path PathFinder::get_path(type::idx_t idx) {
+    if (!computation_launch)
+        return {};
+    ProjectionData projection = this->geo_ref.projected_stop_points[idx][mode];
+
+    auto nearest_edge = find_nearest_vertex(projection);
+
+    return get_path(projection, nearest_edge);
+}
+
+void PathFinder::add_custom_projections_to_path(Path& p,
+                                                bool append_to_begin,
+                                                const ProjectionData& projection,
+                                                ProjectionData::Direction d) const {
+    auto item_to_update = [append_to_begin](Path& p) -> PathItem& {
+        return (append_to_begin ? p.path_items.front() : p.path_items.back());
+    };
+    auto add_in_path = [append_to_begin](Path& p, const PathItem& item) {
+        return (append_to_begin ? p.path_items.push_front(item) : p.path_items.push_back(item));
+    };
+
+    Edge start_edge = projection.edge;
+    nt::LineString coords_to_add;
+    /*
+        Cut the projected edge.
+        We have a point p and a direction d.
+        Our chunk is something like
+                 _______/\           target_e
+        start_e /         \__p______/
+        The third parameter tell if we want the geometry before or after p.
+        If the direction is target_e, we want the end, otherwise the start.
+    */
+    if (start_edge.way_idx != nt::invalid_idx) {
+        Way* way = geo_ref.ways[start_edge.way_idx];
+        if (start_edge.geom_idx != nt::invalid_idx) {
+            coords_to_add =
+                type::split_line_at_point(way->geoms[start_edge.geom_idx], projection.projected, d == target_e);
+        }
+    }
+    if (coords_to_add.empty()) {
+        coords_to_add.push_back(projection.projected);
+    }
+
+    auto duration = crow_fly_duration(projection.distances[d]);
+    // we need to update the total length
+    p.duration += duration;
+
+    // we aither add the starting coordinate to the first path item or create a new path item if it was another way
+    nt::idx_t first_way_idx = (p.path_items.empty() ? type::invalid_idx : item_to_update(p).way_idx);
+    if (start_edge.way_idx != first_way_idx || first_way_idx == type::invalid_idx) {
+        // there can be an item with no way, so we will update this item
+        if (!p.path_items.empty() && item_to_update(p).way_idx == type::invalid_idx) {
+            item_to_update(p).way_idx = start_edge.way_idx;
+            item_to_update(p).duration += duration;
+        } else {
+            PathItem item;
+            item.way_idx = start_edge.way_idx;
+            item.duration = duration;
+
+            if (!p.path_items.empty()) {
+                // still complexifying stuff... TODO: simplify this
```

Comment:
Maybe start with extracting the switch case into a separate function :)
And if you're not familiar with the SOLID principles, have a look at the first one - https://en.wikipedia.org/wiki/SOLID
TL;DR -
- a class has one responsibility
- a function does one thing

Reply:
Is `get_transportation_mode_item_to_update` for the function containing the switch good enough for you ?